GeometryCore/src/LinearAlgebra.cpp

```diff
@@ -4,29 +4,114 @@
+        }
+    }
+
+    FPoint2D::coord distanceBetweenLineSegments(const LineSegment2D& line1, const LineSegment2D& line2) {
+        auto check_projection = [](const FPoint2D& p, const LineSegment2D& line, FPoint2D::coord& minimal_distance) { minimal_distance = std::min(distanceFromPointToLine(p, line), minimal_distance); };
+
+        FPoint2D::coord min_dis = std::min(std::min(distance(line1.a, line2.a), distance(line1.a, line2.b)), std::min(distance(line1.b, line2.a), distance(line1.b, line2.b)));
+        check_projection(line2.a, line1, min_dis);
+        check_projection(line2.b, line1, min_dis);
+        check_projection(line1.a, line2, min_dis);
+        check_projection(line1.b, line2, min_dis);
+
+        return min_dis;
     }
 
     // Return true if poly and line are colliding with each other.
     // Note: consider collision even if one of line end lies exactly on the polygon contour
-    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line)
-    {
-        return {};
+    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line) {
+        auto convex = poly.convexHull();
+        std::vector<bool> line_collide_with_polygon_side;
+        for (int i = 0; i < convex.size(); ++i) {
+                LineSegment2D lie(convex[i], convex[(i + 1) % convex.size()]);
+                line_collide_with_polygon_side.push_back(lineSegmentsCollide(line, lie));
+        }
+        if (std::any_of(line_collide_with_polygon_side.begin(), line_collide_with_polygon_side.end(), [](bool k) {return k; })) {
+            return true;
+        }
+        return false;
+    }
+
+    //Return true if polygon is close to line (closer than given gap)
+    bool polygonAndLineSegmentAreAdjacent(const Polygon& poly, const LineSegment2D& line, const FPoint2D::coord gap) {
+        auto convex = poly.convexHull();
+        std::vector<FPoint2D::coord> distance_from_line_to_polygon_side;
+        #pragma omp parallel for
+        for (int i = 0; i < convex.size(); ++i) {
+                LineSegment2D lie(convex[i], convex[(i + 1) % convex.size()]);
+                #pragma omp atomic
+                distance_from_line_to_polygon_side.push_back(distanceBetweenLineSegments(line, lie));
+        }
+        FPoint2D::coord min_distance = distance_from_line_to_polygon_side[0];
+        for (const auto& i : distance_from_line_to_polygon_side) {
+            min_distance = std::min(min_distance, i);
+        }
+        return min_distance <= gap;
     }
 
     // Return true if poly and poly2 are close to each other (closer than the provided gap).
     // Note: test vertices of one polygon to see if they are closer than the gap to any lines in other poly.
-    bool polygonsAreAdjacent(const Polygon& poly, const Polygon& poly2, const FPoint2D::coord gap)
-    {
-        return {};
+    bool polygonsAreAdjacent(const Polygon& poly, const Polygon& poly2, const FPoint2D::coord gap) {
```

Comment:
Please see my comments to previous polygonAndLineSegmentAreAdjacent() function. My questions/suggestions are the same